### PR TITLE
Ranked projects deep analysis

### PR DESCRIPTION
**Pull request by ryaneveson**
## 📝 Description
This update integrates the deep code analyzer into the project ranking system that i built. Along with this it normalizes all results to a 0–100 score, rather than having random numbers that the user wouldn't understand. It adds the necessary imports, introduces a _perform_deep_code_analysis() helper for analyzing file contents from the database, and updates calculate_project_score(). The new analysis contributes scores across OOP principles, data structures, complexity awareness, optimization techniques, code quality, and additional strengths (this is everything that Sami implemented over reading break)

You are able to see this work by ranking a new project that you upload, and seeing the score that it gives. along with this the summary will tell you what the project contains.

> Please include a summary of the change and which issue is fixed.  
> Include relevant motivation and context.  
> List any dependencies that are required for this change.

**Closes:** # This does not close an issue as this was not an issue that needed to be solved. I spoke with Dr Hui about what i needed to do and these are changes that she suggested to be made. It is addition to a issue that i have already closed.
---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [*] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation added/updated
- [*] ✅ Test added/updated
- [ ] ♻️ Refactoring
- [*] ⚡ Performance improvement

---

## 🧪 Testing

> Please describe how you tested this PR (both manually and with tests).  
> Provide instructions so we can reproduce.
The tests cover all deep analysis scoring components together, including OOP principles, data-structure usage, complexity awareness, optimization patterns, and overall code quality. They verify that projects receive higher scores when deep analysis is enabled and that all results are correctly normalized to the 0–100 range. Additional tests validate edge cases, such as empty analysis results, missing files, exceptions during analysis, and proper capping of maximum scores.

- [ ] Test A
- [ ] Test B

---

## ✓ Checklist

- [*] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [*] 💬 I have commented my code where needed
- [*] 📖 I have made corresponding changes to the documentation
- [*] ⚠️ My changes generate no new warnings
- [*] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [*] 🔗 Any dependent changes have been merged and published in downstream modules
- [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

> If applicable, add screenshots to help explain your changes

<details>
<summary>Click to expand screenshots</summary>

</details>

### CODE REVIEW

**Review by RavenHu001 — APPROVED** [review]
Codes looks good and works fine too.  And to the ranking codes, I think set the max weight scale for each attribute of the code is a good idea, but maybe show the total grade and each  attribute's grade together will be more clear for user.

File: src/analysis/project_ranking.py [hunk 1]
```diff
@@ -50,32 +84,96 @@ def calculate_project_score(analysis_data: Dict[str, Any]) -> float:
...
+            quality_summary = deep_analysis.get("code_quality_summary", {})
+            avg_quality = quality_summary.get("average_quality_score", 0)
+            if avg_quality > 0:
+                # Normalize quality score (0-100) to 0-15 points
+                deep_analysis_score += min(avg_quality * 0.15, 15.0)
+            
+            # Strengths bonus (max 5 points)
+            strengths = quality_summary.get("strengths", [])
+            deep_analysis_score += min(len(strengths) * 1.0, 5.0)
+
```
Comment (RavenHu001) [inline src/analysis/project_ranking.py]: I think we could show these scores after the total score.

**Comment by ryaneveson** [comment]
Optional is imported in the code, i just did not have to add it as it was already imported. I will include a screen shot to show you. 
image
